File: frontend/public/js/complaints.js
// Complaints Management System for MFB Dashboard
// Using Backend API with localStorage fallback

const Complaints = {
  complaints: [],
  config: null,
  useBackend: true, // Toggle for backend API usage
  apiBase: null, // Will be set from config

  // Initialize complaints system
  async init() {
    try {
      // Use absolute path for config
      const configResponse = await fetch('/data/config.json');
      if (!configResponse.ok) throw new Error('Failed to load config');
      this.config = await configResponse.json();
      
      // Use same origin to avoid CORS issues
      // Always use trailing slash to avoid FastAPI redirect issues
      this.apiBase = '/api/complaints';
      
      // Try to load from backend first
      if (this.useBackend) {
        try {
          // Always include trailing slash to avoid redirect
          const response = await fetch(this.apiBase + '/');
          if (response.ok) {
            this.complaints = await response.json();
            console.log('Loaded complaints from backend:', this.complaints.length);
            return true;
          }
        } catch (err) {
          console.warn('Backend unavailable, falling back to localStorage', err);
        }
      }
      
      // Fallback to localStorage
      this.complaints = Utils.storage.get('mfb_complaints', []);
      return true;
    } catch (error) {
      console.error('Failed to initialize complaints:', error);
      return false;
    }
  },

  // Initialize from role pages (deprecated - use init instead)
  async initFromRole() {
    return this.init();
  },

  // Get all complaints
  getAll() {
    return this.complaints.sort((a, b) => new Date(b.createdAt) - new Date(a.createdAt));
  },

  // Get complaint by ID
  getById(id) {
    return this.complaints.find(c => c.id === id);
  },

  // Create new complaint
  async create(complaintData) {
    const user = Auth.getCurrentUser();
    const now = new Date().toISOString();
    
    if (this.useBackend) {
      try {
        const response = await fetch(this.apiBase + '/', {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({
            ...complaintData,
            createdBy: user ? user.id : 'System',
            createdByName: user ? user.name : 'System'
          })
        });
        
        if (response.ok) {
          const complaint = await response.json();
          this.complaints.unshift(complaint);
          return complaint;
        }
      } catch (err) {
        console.error('Backend create failed:', err);
      }
    }
    
    // Fallback to localStorage
    const complaint = {
      id: 'CMP' + Date.now().toString(36).toUpperCase(),
      ticketNumber: 'TKT-' + String(this.complaints.length + 1).padStart(6, '0'),
      ...complaintData,
      status: 'Open',
      priority: complaintData.priority || 'Medium',
      createdAt: now,
      updatedAt: now,
      createdBy: user ? user.id : 'System',
      createdByName: user ? user.name : 'System',
      assignedTo: null,
      assignedToName: null,
      escalatedTo: null,
      escalationReason: null,
      resolvedAt: null,
      resolution: null,
      slaDeadline: this.calculateSLADeadline(now, complaintData.priority || 'Medium'),
      history: [{
        action: 'Created',
        timestamp: now,
        user: user ? user.name : 'System',
        details: 'Complaint created'
      }]
    };

    this.complaints.push(complaint);
    this.save();
    return complaint;
  },

  // Update complaint
  async update(id, updates) {
    const index = this.complaints.findIndex(c => c.id === id);
    if (index === -1) return null;

    if (this.useBackend) {
      try {
        const response = await fetch(`${this.apiBase}/${id}`, {
          method: 'PATCH',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify(updates)
        });
        
        if (response.ok) {
          const updated = await response.json();
          this.complaints[index] = updated;
          return updated;
        }
      } catch (err) {
        console.error('Backend update failed:', err);
      }
    }

    // Fallback
    const user = Auth.getCurrentUser();
    const now = new Date().toISOString();

    const historyEntry = {
      action: 'Updated',
      timestamp: now,
      user: user ? user.name : 'System',
      details: Object.keys(updates).join(', ') + ' updated'
    };

    this.complaints[index] = {
      ...this.complaints[index],
      ...updates,
      updatedAt: now,
      history: [...this.complaints[index].history, historyEntry]
    };

    this.save();
    return this.complaints[index];
  },

  // Assign complaint
  async assign(id, staffId, staffName) {
    const complaint = this.getById(id);
    if (!complaint) return null;

    if (this.useBackend) {
      try {
        const user = Auth.getCurrentUser();
        const response = await fetch(`${this.apiBase}/${id}/action`, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({
            action: 'Assign',
            user: user ? user.id : 'System',
            userName: user ? user.name : 'System',
            assignedTo: staffId,
            assignedToName: staffName
          })
        });
        
        if (response.ok) {
          const updated = await response.json();
          const index = this.complaints.findIndex(c => c.id === id);
          if (index !== -1) this.complaints[index] = updated;
          return updated;
        }
      } catch (err) {
        console.error('Backend assign failed:', err);
      }
    }

    // Fallback
    const user = Auth.getCurrentUser();
    const now = new Date().toISOString();

    complaint.assignedTo = staffId;
    complaint.assignedToName = staffName;
    complaint.status = complaint.status === 'Open' ? 'In Progress' : complaint.status;
    complaint.updatedAt = now;
    complaint.history.push({
      action: 'Assigned',
      timestamp: now,
      user: user ? user.name : 'System',
      details: `Assigned to ${staffName}`
    });

    this.save();
    return complaint;
  },

  // Escalate complaint
  async escalate(id, escalatedTo, reason) {
    const complaint = this.getById(id);
    if (!complaint) return null;

    if (this.useBackend) {
      try {
        const user = Auth.getCurrentUser();
        const response = await fetch(`${this.apiBase}/${id}/action`, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({
            action: 'Escalate',
            user: user ? user.id : 'System',
            userName: user ? user.name : 'System',
            escalatedTo: escalatedTo,
            escalationReason: reason
          })
        });
        
        if (response.ok) {
          const updated = await response.json();
          const index = this.complaints.findIndex(c => c.id === id);
          if (index !== -1) this.complaints[index] = updated;
          return updated;
        }
      } catch (err) {
        console.error('Backend escalate failed:', err);
      }
    }

    // Fallback
    const user = Auth.getCurrentUser();
    const now = new Date().toISOString();

    complaint.status = 'Escalated';
    complaint.escalatedTo = escalatedTo;
    complaint.escalationReason = reason;
    complaint.updatedAt = now;
    complaint.history.push({
      action: 'Escalated',
      timestamp: now,
      user: user ? user.name : 'System',
      details: `Escalated to ${escalatedTo}. Reason: ${reason}`
    });

    this.save();
    return complaint;
  },

  // Resolve complaint
  async resolve(id, resolution) {
    const complaint = this.getById(id);
    if (!complaint) return null;

    if (this.useBackend) {
      try {
        const user = Auth.getCurrentUser();
        const response = await fetch(`${this.apiBase}/${id}/action`, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({
            action: 'Resolve',
            user: user ? user.id : 'System',
            userName: user ? user.name : 'System',
            resolution: resolution
          })
        });
        
        if (response.ok) {
          const updated = await response.json();
          const index = this.complaints.findIndex(c => c.id === id);
          if (index !== -1) this.complaints[index] = updated;
          return updated;
        }
      } catch (err) {
        console.error('Backend resolve failed:', err);
      }
    }

    // Fallback
    const user = Auth.getCurrentUser();
    const now = new Date().toISOString();

    complaint.status = 'Resolved';
    complaint.resolution = resolution;
    complaint.resolvedAt = now;
    complaint.updatedAt = now;
    complaint.history.push({
      action: 'Resolved',
      timestamp: now,
      user: user ? user.name : 'System',
      details: `Resolved: ${resolution}`
    });

    this.save();
    return complaint;
  },

  // Close complaint
  async close(id) {
    const complaint = this.getById(id);
    if (!complaint) return null;

    if (this.useBackend) {
      try {
        const user = Auth.getCurrentUser();
        const response = await fetch(`${this.apiBase}/${id}/action`, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({
            action: 'Close',
            user: user ? user.id : 'System',
            userName: user ? user.name : 'System'
          })
        });
        
        if (response.ok) {
          const updated = await response.json();
          const index = this.complaints.findIndex(c => c.id === id);
          if (index !== -1) this.complaints[index] = updated;
          return updated;
        }
      } catch (err) {
        console.error('Backend close failed:', err);
      }
    }

    // Fallback
    const user = Auth.getCurrentUser();
    const now = new Date().toISOString();

    complaint.status = 'Closed';
    complaint.updatedAt = now;
    complaint.history.push({
      action: 'Closed',
      timestamp: now,
      user: user ? user.name : 'System',
      details: 'Complaint closed'
    });

    this.save();
    return complaint;
  },

  // Reopen complaint
  async reopen(id, reason) {
    const complaint = this.getById(id);
    if (!complaint) return null;

    if (this.useBackend) {
      try {
        const user = Auth.getCurrentUser();
        const response = await fetch(`${this.apiBase}/${id}/action`, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({
            action: 'Reopen',
            user: user ? user.id : 'System',
            userName: user ? user.name : 'System',
            details: reason
          })
        });
        
        if (response.ok) {
          const updated = await response.json();
          const index = this.complaints.findIndex(c => c.id === id);
          if (index !== -1) this.complaints[index] = updated;
          return updated;
        }
      } catch (err) {
        console.error('Backend reopen failed:', err);
      }
    }

    // Fallback
    const user = Auth.getCurrentUser();
    const now = new Date().toISOString();

    complaint.status = 'Open';
    complaint.resolvedAt = null;
    complaint.resolution = null;
    complaint.updatedAt = now;
    complaint.slaDeadline = this.calculateSLADeadline(now, complaint.priority);
    complaint.history.push({
      action: 'Reopened',
      timestamp: now,
      user: user ? user.name : 'System',
      details: `Reopened: ${reason}`
    });

    this.save();
    return complaint;
  },

  // Add note to complaint
  async addNote(id, note) {
    const complaint = this.getById(id);
    if (!complaint) return null;

    if (this.useBackend) {
      try {
        const user = Auth.getCurrentUser();
        const response = await fetch(`${this.apiBase}/${id}/action`, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({
            action: 'Note Added',
            user: user ? user.id : 'System',
            userName: user ? user.name : 'System',
            details: note
          })
        });
        
        if (response.ok) {
          const updated = await response.json();
          const index = this.complaints.findIndex(c => c.id === id);
          if (index !== -1) this.complaints[index] = updated;
          return updated;
        }
      } catch (err) {
        console.error('Backend addNote failed:', err);
      }
    }

    // Fallback
    const user = Auth.getCurrentUser();
    const now = new Date().toISOString();

    complaint.updatedAt = now;
    complaint.history.push({
      action: 'Note Added',
      timestamp: now,
      user: user ? user.name : 'System',
      details: note
    });

    this.save();
    return complaint;
  },

  // Calculate SLA deadline
  calculateSLADeadline(createdAt, priority) {
    const slaHours = this.config?.slaHours || {
      Low: 72,
      Medium: 48,
      High: 24,
      Critical: 4
    };
    const hours = slaHours[priority] || 48;
    const deadline = new Date(createdAt);
    deadline.setHours(deadline.getHours() + hours);
    return deadline.toISOString();
  },

  // Get SLA status
  getSLAStatus(complaint) {
    if (complaint.status === 'Closed' || complaint.status === 'Resolved') {
      const resolved = new Date(complaint.resolvedAt || complaint.updatedAt);
      const deadline = new Date(complaint.slaDeadline);
      return resolved <= deadline ? 'Met' : 'Breached';
    }

    const now = new Date();
    const deadline = new Date(complaint.slaDeadline);
    const remaining = deadline - now;
    const hoursRemaining = remaining / (1000 * 60 * 60);

    if (remaining < 0) return 'Breached';
    if (hoursRemaining < 2) return 'Critical';
    if (hoursRemaining < 8) return 'Warning';
    return 'On Track';
  },

  // Get complaints by status
  getByStatus(status) {
    return this.complaints.filter(c => c.status === status);
  },

  // Get complaints by assigned user
  getByAssignee(userId) {
    return this.complaints.filter(c => c.assignedTo === userId);
  },

  // Get complaints by customer
  getByCustomer(customerId) {
    return this.complaints.filter(c => c.customerId === customerId);
  },

  // Get complaints analytics
  getAnalytics() {
    const total = this.complaints.length;
    const byStatus = Utils.countBy(this.complaints, 'status');
    const byCategory = Utils.countBy(this.complaints, 'category');
    const byPriority = Utils.countBy(this.complaints, 'priority');

    // SLA metrics
    const closed = this.complaints.filter(c => c.status === 'Closed' || c.status === 'Resolved');
    const slaMet = closed.filter(c => this.getSLAStatus(c) === 'Met').length;
    const slaBreached = closed.filter(c => this.getSLAStatus(c) === 'Breached').length;

    // Active complaints at risk
    const active = this.complaints.filter(c => !['Closed', 'Resolved'].includes(c.status));
    const atRisk = active.filter(c => ['Critical', 'Warning', 'Breached'].includes(this.getSLAStatus(c)));

    // Average resolution time
    const resolvedComplaints = closed.filter(c => c.resolvedAt);
    let avgResolutionTime = 0;
    if (resolvedComplaints.length > 0) {
      const totalTime = resolvedComplaints.reduce((sum, c) => {
        return sum + (new Date(c.resolvedAt) - new Date(c.createdAt));
      }, 0);
      avgResolutionTime = totalTime / resolvedComplaints.length / (1000 * 60 * 60); // In hours
    }

    // Daily trends
    const dailyComplaints = {};
    this.complaints.forEach(c => {
      const date = c.createdAt.split('T')[0];
      dailyComplaints[date] = (dailyComplaints[date] || 0) + 1;
    });

    return {
      total,
      open: byStatus['Open'] || 0,
      inProgress: byStatus['In Progress'] || 0,
      escalated: byStatus['Escalated'] || 0,
      resolved: byStatus['Resolved'] || 0,
      closed: byStatus['Closed'] || 0,
      byCategory,
      byPriority,
      slaMet,
      slaBreached,
      slaRate: closed.length ? (slaMet / closed.length) * 100 : 0,
      atRiskCount: atRisk.length,
      avgResolutionTime,
      dailyTrends: Object.entries(dailyComplaints)
        .map(([date, count]) => ({ date, count }))
        .sort((a, b) => a.date.localeCompare(b.date))
    };
  },

  // Save to localStorage (fallback)
  save() {
    Utils.storage.set('mfb_complaints', this.complaints);
  },

  // Search complaints
  search(query) {
    const q = query.toLowerCase();
    return this.complaints.filter(c => 
      c.ticketNumber?.toLowerCase().includes(q) ||
      c.customerName?.toLowerCase().includes(q) ||
      c.customerPhone?.toLowerCase().includes(q) ||
      c.customerAccountNumber?.toLowerCase().includes(q) ||
      c.description?.toLowerCase().includes(q) ||
      c.category?.toLowerCase().includes(q)
    );
  },

  // Filter complaints
  filter(filters) {
    return this.complaints.filter(c => {
      if (filters.status && c.status !== filters.status) return false;
      if (filters.category && c.category !== filters.category) return false;
      if (filters.priority && c.priority !== filters.priority) return false;
      if (filters.assignedTo && c.assignedTo !== filters.assignedTo) return false;
      if (filters.dateFrom && new Date(c.createdAt) < new Date(filters.dateFrom)) return false;
      if (filters.dateTo && new Date(c.createdAt) > new Date(filters.dateTo)) return false;
      return true;
    });
  }
};

// Export
if (typeof module !== 'undefined' && module.exports) {
  module.exports = Complaints;
}

File: backend/routes/complaints.py
from fastapi import APIRouter, HTTPException, status
from pydantic import BaseModel, Field
from typing import List, Optional, Dict, Any
from datetime import datetime, timezone, timedelta
import uuid

router = APIRouter(prefix="/complaints", tags=["complaints"])

# In-memory database reference (will be set from server.py)
db = None

def set_db(database):
    global db
    db = database

# Models - Using simple dict-like structure for flexibility
class ComplaintCreate(BaseModel):
    customerName: str
    customerPhone: Optional[str] = None
    customerAccountNumber: Optional[str] = None
    customerId: Optional[str] = None
    category: str
    priority: str = "Medium"
    channel: str = "Walk-in"
    subject: Optional[str] = None
    description: Optional[str] = None
    relatedTransaction: Optional[str] = None
    assignedTo: Optional[str] = None
    assignedToName: Optional[str] = None
    createdBy: Optional[str] = None
    createdByName: Optional[str] = None

class ComplaintUpdate(BaseModel):
    status: Optional[str] = None
    priority: Optional[str] = None
    assignedTo: Optional[str] = None
    assignedToName: Optional[str] = None
    resolution: Optional[str] = None
    escalatedTo: Optional[str] = None
    escalationReason: Optional[str] = None

class ActionRequest(BaseModel):
    action: str
    user: str
    userName: str
    details: Optional[str] = None
    assignedTo: Optional[str] = None
    assignedToName: Optional[str] = None
    resolution: Optional[str] = None
    escalatedTo: Optional[str] = None
    escalationReason: Optional[str] = None

# Helper functions
def calculate_sla_deadline(priority: str) -> str:
    """Calculate SLA deadline based on priority"""
    sla_hours = {
        "Low": 72,
        "Medium": 48,
        "High": 24,
        "Critical": 4
    }
    hours = sla_hours.get(priority, 48)
    deadline = datetime.now(timezone.utc) + timedelta(hours=hours)
    return deadline.isoformat()

# Routes
@router.get("/")
async def get_all_complaints():
    """Get all complaints"""
    complaints = await db.complaints.find({}, {"_id": 0}).sort("createdAt", -1).to_list(1000)
    return complaints

@router.get("/analytics/summary")
async def get_analytics():
    """Get complaint analytics summary"""
    total = await db.complaints.count_documents({})
    
    pipeline = [
        {"$group": {"_id": "$status", "count": {"$sum": 1}}}
    ]
    status_counts = await db.complaints.aggregate(pipeline).to_list(100)
    by_status = {item["_id"]: item["count"] for item in status_counts if item["_id"]}
    
    pipeline = [
        {"$group": {"_id": "$category", "count": {"$sum": 1}}}
    ]
    category_counts = await db.complaints.aggregate(pipeline).to_list(100)
    by_category = {item["_id"]: item["count"] for item in category_counts if item["_id"]}
    
    pipeline = [
        {"$group": {"_id": "$priority", "count": {"$sum": 1}}}
    ]
    priority_counts = await db.complaints.aggregate(pipeline).to_list(100)
    by_priority = {item["_id"]: item["count"] for item in priority_counts if item["_id"]}
    
    return {
        "total": total,
        "open": by_status.get("Open", 0),
        "inProgress": by_status.get("In Progress", 0),
        "escalated": by_status.get("Escalated", 0),
        "resolved": by_status.get("Resolved", 0),
        "closed": by_status.get("Closed", 0),
        "byCategory": by_category,
        "byPriority": by_priority,
        "byStatus": by_status
    }

@router.get("/status/{status}")
async def get_complaints_by_status(status: str):
    """Get complaints by status"""
    complaints = await db.complaints.find(
        {"status": status}, 
        {"_id": 0}
    ).sort("createdAt", -1).to_list(1000)
    return complaints

@router.get("/assignee/{user_id}")
async def get_complaints_by_assignee(user_id: str):
    """Get complaints assigned to a specific user"""
    complaints = await db.complaints.find(
        {"assignedTo": user_id}, 
        {"_id": 0}
    ).sort("createdAt", -1).to_list(1000)
    return complaints

@router.get("/{complaint_id}")
async def get_complaint(complaint_id: str):
    """Get a single complaint by ID"""
    complaint = await db.complaints.find_one({"id": complaint_id}, {"_id": 0})
    if not complaint:
        raise HTTPException(status_code=404, detail="Complaint not found")
    return complaint

@router.post("/", status_code=status.HTTP_201_CREATED)
async def create_complaint(complaint_data: ComplaintCreate):
    """Create a new complaint"""
    now = datetime.now(timezone.utc).isoformat()
    
    # Get next ticket number
    count = await db.complaints.count_documents({})
    ticket_number = f"TKT-{str(count + 1).zfill(6)}"
    
    complaint = {
        "id": f"CMP{uuid.uuid4().hex[:12].upper()}",
        "ticketNumber": ticket_number,
        "customerName": complaint_data.customerName,
        "customerPhone": complaint_data.customerPhone,
        "customerAccountNumber": complaint_data.customerAccountNumber,
        "customerId": complaint_data.customerId,
        "category": complaint_data.category,
        "priority": complaint_data.priority,
        "channel": complaint_data.channel,
        "subject": complaint_data.subject,
        "description": complaint_data.description,
        "relatedTransaction": complaint_data.relatedTransaction,
        "status": "Open",
        "assignedTo": complaint_data.assignedTo,
        "assignedToName": complaint_data.assignedToName,
        "escalatedTo": None,
        "escalationReason": None,
        "resolution": None,
        "resolvedAt": None,
        "slaDeadline": calculate_sla_deadline(complaint_data.priority),
        "createdAt": now,
        "updatedAt": now,
        "createdBy": complaint_data.createdBy,
        "createdByName": complaint_data.createdByName,
        "history": [{
            "action": "Created",
            "timestamp": now,
            "user": complaint_data.createdByName or "System",
            "details": "Complaint created"
        }]
    }
    
    await db.complaints.insert_one(complaint)
    
    # Remove MongoDB _id before returning
    complaint.pop("_id", None)
    return complaint

@router.patch("/{complaint_id}")
async def update_complaint(complaint_id: str, updates: ComplaintUpdate):
    """Update a complaint"""
    complaint = await db.complaints.find_one({"id": complaint_id})
    if not complaint:
        raise HTTPException(status_code=404, detail="Complaint not found")
    
    update_dict = {"updatedAt": datetime.now(timezone.utc).isoformat()}
    
    if updates.status:
        update_dict["status"] = updates.status
    if updates.priority:
        update_dict["priority"] = updates.priority
        update_dict["slaDeadline"] = calculate_sla_deadline(updates.priority)
    if updates.assignedTo:
        update_dict["assignedTo"] = updates.assignedTo
    if updates.assignedToName:
        update_dict["assignedToName"] = updates.assignedToName
    if updates.resolution:
        update_dict["resolution"] = updates.resolution
    if updates.escalatedTo:
        update_dict["escalatedTo"] = updates.escalatedTo
    if updates.escalationReason:
        update_dict["escalationReason"] = updates.escalationReason
    
    await db.complaints.update_one(
        {"id": complaint_id},
        {"$set": update_dict}
    )
    
    updated = await db.complaints.find_one({"id": complaint_id}, {"_id": 0})
    return updated

@router.post("/{complaint_id}/action")
async def perform_action(complaint_id: str, action: ActionRequest):
    """Perform an action on a complaint (assign, escalate, resolve, close, reopen, add note)"""
    complaint = await db.complaints.find_one({"id": complaint_id})
    if not complaint:
        raise HTTPException(status_code=404, detail="Complaint not found")
    
    now = datetime.now(timezone.utc).isoformat()
    update_dict = {"updatedAt": now}
    
    history_entry = {
        "action": action.action,
        "timestamp": now,
        "user": action.userName,
        "details": action.details or f"{action.action} performed"
    }
    
    if action.action == "Assign":
        update_dict["assignedTo"] = action.assignedTo
        update_dict["assignedToName"] = action.assignedToName
        if complaint.get("status") == "Open":
            update_dict["status"] = "In Progress"
        history_entry["details"] = f"Assigned to {action.assignedToName}"
    
    elif action.action == "Start Working":
        update_dict["status"] = "In Progress"
        update_dict["assignedTo"] = action.user
        update_dict["assignedToName"] = action.userName
        history_entry["details"] = f"{action.userName} started working on this complaint"
    
    elif action.action == "Escalate":
        update_dict["status"] = "Escalated"
        update_dict["escalatedTo"] = action.escalatedTo
        update_dict["escalationReason"] = action.escalationReason
        history_entry["details"] = f"Escalated to {action.escalatedTo}. Reason: {action.escalationReason}"
    
    elif action.action == "Resolve":
        update_dict["status"] = "Resolved"
        update_dict["resolution"] = action.resolution
        update_dict["resolvedAt"] = now
        history_entry["details"] = f"Resolved: {action.resolution}"
    
    elif action.action == "Close":
        update_dict["status"] = "Closed"
        history_entry["details"] = "Complaint closed"
    
    elif action.action == "Reopen":
        update_dict["status"] = "Open"
        update_dict["resolvedAt"] = None
        update_dict["resolution"] = None
        update_dict["slaDeadline"] = calculate_sla_deadline(complaint.get("priority", "Medium"))
        history_entry["details"] = f"Reopened: {action.details}"
    
    elif action.action == "Note Added":
        history_entry["details"] = action.details
    
    await db.complaints.update_one(
        {"id": complaint_id},
        {
            "$set": update_dict,
            "$push": {"history": history_entry}
        }
    )
    
    updated = await db.complaints.find_one({"id": complaint_id}, {"_id": 0})
    return updated

@router.delete("/{complaint_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_complaint(complaint_id: str):
    """Delete a complaint"""
    result = await db.complaints.delete_one({"id": complaint_id})
    if result.deleted_count == 0:
        raise HTTPException(status_code=404, detail="Complaint not found")
    return None
